### PR TITLE
#80 Added JointChain that allows visualizing skeleton of a SkinningMesh

### PR DESCRIPTION
**Pull request by jperedadnr**
To test, run sampler, select Maya sample, go to 3D settings, and enable skeleton. Then play the animation.

### CODE REVIEW

File: FXyz-Core/src/main/java/org/fxyz3d/shapes/polygon/SkinningMesh.java [hunk 1]
```diff
@@ -203,4 +207,38 @@ public void update() {
...
+    }
+    public final boolean isShowSkeleton() {
+       return showSkeleton.get();
+    }
+    public final void setShowSkeleton(boolean value) {
+        showSkeleton.set(value);
+    }
+
+    private void addJointChain(Joint joint) {
+        joint.getChildren().removeIf(JointChain.class::isInstance);
```
Comment (nlisker) [inline FXyz-Core/src/main/java/org/fxyz3d/shapes/polygon/SkinningMesh.java]: Any reason not to use `removeJointChain` here to make it easy to see that adding a chain first removes the current one?

**Comment by jperedadnr** [comment]
I can't reproduce the stuttering in my Mac. The line you mention updates the frustum mesh, so it probably is quite time and resource consuming. Not sure how we could improve this?

File: FXyz-Core/src/main/java/org/fxyz3d/utils/geom/JointChain.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    private void updateChain(Joint joint) {
+        Point3D scaled = getJointLocation(joint).multiply(1d / scale);
+        origin.setTranslateX(scaled.getX());
+        origin.setTranslateY(scaled.getY());
+        origin.setTranslateZ(scaled.getZ());
+        bone.setAxisOrigin(org.fxyz3d.geometry.Point3D.convertFromJavaFXPoint3D(scaled));
+        end.setTranslateX(bone.getAxisEnd().x);
+        end.setTranslateY(bone.getAxisEnd().y);
+        end.setTranslateZ(bone.getAxisEnd().z);
+
```
Comment (nlisker) [inline FXyz-Core/src/main/java/org/fxyz3d/utils/geom/JointChain.java]: Remove empty line.

**Comment by nlisker** [comment]
I'm on Win10. I assume I will have to do performance analysis and see what the bottleneck is.

@AlmasB Do you get this stuttering too?

**Comment by AlmasB** [comment]
Sorry I'm busy this week, will check at the weekend

**Comment by nlisker** [comment]
Yes, looks like the problem is that updating the `FrustumMesh` calls the expensive `createFrustum`:

image

I'm not familiar with this class so I don't know why updating a property must create a new one. Also, why does the bone have to be a `FrustumMesh`, is a `Cylinder` not enough?

**Comment by nlisker** [comment]
Seems like the issue persists (maybe not as badly though). I'll have to do profiling again, but only later today.

I wonder if this is somehow related to #52. Can you try to reproduce that one?

The hardware I'm on is i5-6500 and RX470 4GB if it helps with the comparison.

**Comment by nlisker** [comment]
Looks like the issue is with `append` inside `getJointLocation`. If I do not append anything,

    Affine a = new Affine();
    for (int i = 0; i < 0; i++) {
        a.append(joint.getTransforms().get(i));
    }

there is no performance issue. If I change to `i < 1` it's stuttering. I'm looking into circumventing that chain, but the proper thing to do might be to create an MCVE and see if it's a bug in JavaFX.

**Comment by nlisker** [comment]
By the way, I think you can merge the Cylinder revision for now if you want (I approved the PR since the performance issue is local). If you want to show the direction of the joint, maybe use a gradient texture?

Also, it might be a good idea to look into adding a cone, pyramid and maybe other basic shapes directly to JavaFX.

Did you have a chance to test #52? Could help with a direction.

**Comment by jperedadnr** [comment]
I've merged it now as is. We shouldn't forget about the `append` issue though.

I've posted this [video](https://twitter.com/JPeredaDnr/status/1098912324889571328) btw, doing some tweaks to include a filled model and a wireframe model to view the skeleton...

**Pull request merged**